### PR TITLE
Filter out TiFlash stores when checking upStores before TiKV member scaling in

### PR DESCRIPTION
**Pull request by handlerww**
### What problem does this PR solve? 
Filter out TiFlash stores when checking working TiKV Stores before TiKV member scaling in.
### What is changed and how does it work?
When controller get storesinfo from PD, it didn't tell apart the TiFlash stores or TiKV stores. It may cause error when TiFlash is deployed and TiKV is scaling in.
### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)
image
image

Successfully set off the alert.

Code changes

 - Has Go code change

Side effects

 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
Fix the issue that when TiFlash is deployed, TiKV can be scaled in even if there are only 3 UP stores in the cluster
```

### CODE REVIEW

File: pkg/manager/member/tikv_scaler.go [hunk 1]
```diff
@@ -247,8 +248,11 @@ func (s *tikvScaler) preCheckUpStores(tc *v1alpha1.TidbCluster, podName string)
 	if err != nil {
 		return false, fmt.Errorf("failed to get stores info in TidbCluster %s/%s", tc.GetNamespace(), tc.GetName())
 	}
+	// filter out TiFlash
 	for _, store := range storesInfo.Stores {
-		if store.Store != nil && store.Store.StateName == v1alpha1.TiKVStateUp {
+		storePodName := strings.Split(store.Store.Address, ".")[0]
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_scaler.go:253]: Check `store.Store != nil` first.

File: pkg/manager/member/tikv_scaler.go [hunk 2]
```diff
@@ -247,8 +248,11 @@ func (s *tikvScaler) preCheckUpStores(tc *v1alpha1.TidbCluster, podName string)
...
+		podKind := strings.Split(storePodName, "-")
+		if store.Store != nil && store.Store.StateName == v1alpha1.TiKVStateUp && podKind[len(podKind)-2] == "tikv" {
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_scaler.go:255]: Use `label.TiKVLabelVal` for `tikv`.

Comment (DanielZhangQD) [reply pkg/manager/member/tikv_scaler.go:255]: Please also add UT to cover the code change.

File: pkg/manager/member/tikv_scaler_test.go [hunk 3]
```diff
@@ -221,6 +222,9 @@ func TestTiKVScalerScaleIn(t *testing.T) {
 				store := &pdapi.StoreInfo{
 					Store: &pdapi.MetaStore{
 						StateName: v1alpha1.TiKVStateUp,
+						Store: &metapb.Store{
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_scaler_test.go:225]: Should also add tiflash stores to cover your code change?

Comment (handlerww) [reply pkg/manager/member/tikv_scaler_test.go:225]: fixed

**Review by DanielZhangQD — APPROVED** [review]
LGTM

File: pkg/manager/member/tikv_scaler.go [hunk 4]
```diff
@@ -247,9 +248,14 @@ func (s *tikvScaler) preCheckUpStores(tc *v1alpha1.TidbCluster, podName string)
 	if err != nil {
 		return false, fmt.Errorf("failed to get stores info in TidbCluster %s/%s", tc.GetNamespace(), tc.GetName())
 	}
+	// filter out TiFlash
 	for _, store := range storesInfo.Stores {
-		if store.Store != nil && store.Store.StateName == v1alpha1.TiKVStateUp {
-			upNumber++
+		if store.Store != nil {
+			storePodName := strings.Split(store.Store.Address, ".")[0]
+			podKind := strings.Split(storePodName, "-")
```
Comment (lonng) [inline pkg/manager/member/tikv_scaler.go:255]: Do we need to check the `len(podKind)`?

Comment (handlerww) [reply pkg/manager/member/tikv_scaler.go:255]: The pod name is generated in the format "xxx-xxx-tikv(component)-0(ordinals)", the last but one is fixed. Is there any corner case on it?

Comment (handlerww) [reply pkg/manager/member/tikv_scaler.go]: Oh, I see. There might be some instances from outside K8S clusters.

Comment (handlerww) [reply pkg/manager/member/tikv_scaler.go]: Fixed. I have updated UT to cover store named "outsidetikv".

File: pkg/util/util.go [hunk 5]
```diff
@@ -355,6 +355,12 @@ func VolumeClaimTemplate(r corev1.ResourceRequirements, metaName string, storage
 	}
 }
 
+func MatchLabelFromStoreAddress(Address string, label string) bool {
+	storePodName := strings.Split(Address, ".")[0]
+	storeKind := strings.Split(storePodName, "-")
+	return len(storeKind) >= 2 && storeKind[len(storeKind)-2] == label
```
Comment (DanielZhangQD) [inline pkg/util/util.go:361]: If `len(storeKind) < 2`, we should also count them in.

File: pkg/util/util.go [hunk 6]
```diff
@@ -355,6 +356,17 @@ func VolumeClaimTemplate(r corev1.ResourceRequirements, metaName string, storage
 	}
 }
 
+func MatchLabelFromStoreLabels(Labels []*metapb.StoreLabel, componentLabel string) bool {
```
Comment (DanielZhangQD) [inline pkg/util/util.go:359]:
```suggestion
func MatchLabelFromStoreLabels(labels []*metapb.StoreLabel, componentLabel string) bool {
```

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Pull request merged**